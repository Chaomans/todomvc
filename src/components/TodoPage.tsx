import { useState } from "react";
import { Filter, TodoItem } from "./models/models";
import TodoFooter from "./TodoFooter";
import TodoInput from "./TodoInput";
import TodoList from "./TodoList";

const TodoPage = () => {

    const [todoList, setTodoList] = useState<TodoItem[]>([]);

    const filters: Filter[] = [
        {
            name: "All",
            value: "All"
        },
        {
            name: "Active",
            value: "Active"
        },
        {
            name: "Completed",
            value: "Completed"
        }
    ];
    const [activeFilter, setActiveFilter] = useState("All");

    const handleChangeActiveFilter = (newActiveFilter: string) => {
        setActiveFilter(newActiveFilter);
    }

    const handleCreate = (description: string) => {
        setTodoList([
            ...todoList,
            {
                id: crypto.randomUUID(),
                description,
                done: false
            }
        ])
    };

    const handleDelete = (id: string) => {
        setTodoList(todoList.filter((todo) => todo.id !== id));
    };

    const handleChangeDoneValue = (id: string, done: boolean) => {
        const newTodoList = todoList.map((todo) => {
            if (todo.id === id) {
                return ({
                    ...todo,
                    done
                });
            } else {
                return todo;
            }
        });
        setTodoList(newTodoList);
    };

    const handleEditDescription = (id: string, desc: string) => {
        const toEditIndex =
            todoList.findIndex((todo) => todo.id === id) ?? null;
        if (toEditIndex === null) return;
        const newTodoList: TodoItem[] = [];
        todoList.forEach((todo, i) => {
            if (i === toEditIndex) {
                newTodoList.push({
                    ...todoList[toEditIndex],
                    description: desc,
                });
            } else {
                newTodoList.push(todo);
            }
        });
        setTodoList(newTodoList);
    };

    const handleChangeAllDoneValue = () => {
        const alldone = todoList.every((todo) => todo.done);
        const newTodoList: TodoItem[] = [];
        if (alldone) {
            todoList.forEach((todo) =>
                newTodoList.push({
                    ...todo,
                    done: false,
                })
            );
        } else {
            todoList.forEach((todo) =>
                newTodoList.push({
                    ...todo,
                    done: true,
                })
            );
        }
        setTodoList(newTodoList);
    };

    const handleClearCompleted = () => {
        setTodoList(todoList.filter((todo) => !todo.done));
    }

    const filteredTodoList = (filter: string) => {
        if (filter === "All") return todoList;
        if (filter === "Active") {
            return todoList.filter((todo) => !todo.done);
        }
        if(filter === "Completed"){
            return todoList.filter((todo) => todo.done);
        }
        return []
    }

    return (
        <section className="todoapp">
            <TodoInput createTodo={handleCreate} />
            <section className="main">
                <input
                    type="checkbox"
                    id="toggle-all"
                    className="toggle-all"
                    onChange={() => handleChangeAllDoneValue()}
                />
                <label htmlFor="toggle-all">Mark all as completed</label>
                {todoList.length > 0 && (
                    <TodoList 
                        todos={filteredTodoList(activeFilter)}
                        onChangeDoneValue={handleChangeDoneValue}
                        onEditDescription={handleEditDescription}
                        onDelete={handleDelete}
                    />
                )}
            </section>
            {todoList.length > 0 && (
                <TodoFooter
                    remainingTodos={todoList.filter((todo) => !todo.done).length}
                    completedTodos={todoList.filter((todo) => todo.done).length}
                    filters={filters}
                    activeFilter={activeFilter}
                    onChangeActiveFilter={handleChangeActiveFilter}
                    onClearCompleted={handleClearCompleted}
                />
            )}
        </section>
    );
};

export default TodoPage;