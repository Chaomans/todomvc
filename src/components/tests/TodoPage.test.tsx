import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TodoPage from "../TodoPage";


const crypto = require('crypto');

Object.defineProperty(globalThis, 'crypto', {
  value: {
    randomUUID: () => crypto.randomUUID()
  }
});

describe("TodoPage component", () => {
    it("Should render properly", () => {
        const { asFragment } = render(
            <TodoPage/>
        )

        expect(asFragment()).toMatchSnapshot();
    });

    it("Should create a todo", async () => {
        render(
            <TodoPage/>
        )

        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a rainbow{Enter}");
        expect(await screen.findByText("eat a rainbow")).toBeInTheDocument();
        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a burger{Enter}");
        expect(await screen.findByText("eat a burger")).toBeInTheDocument();
        const lists = await screen.findAllByRole("list");
        expect(lists[0].childElementCount).toEqual(2);
    });

    it("Should delete the chosen todo", async () => {
        render(
            <TodoPage/>
        )

        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a rainbow{Enter}");
        expect(await screen.findByText("eat a rainbow")).toBeInTheDocument();
        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a burger{Enter}");
        expect(await screen.findByText("eat a burger")).toBeInTheDocument();
        const lists = await screen.findAllByRole("list");
        expect(lists[0].childElementCount).toEqual(2);
        const buttons = await within(lists[0]).findAllByRole("button");
        await userEvent.click(buttons[0]);
        expect(screen.queryByText("eat a rainbow")).not.toBeInTheDocument();
        expect(lists[0].childElementCount).toEqual(1);
    });

    it("Should show list and footer if 1+ todos", async () => {
        render(
            <TodoPage/>
        )

        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a rainbow{Enter}");
        expect(document.querySelector(".main")).toBeInTheDocument();
        expect(document.querySelector(".footer")).toBeInTheDocument();
    });

    it.each`
    filter          
    ${"All"}      
    ${"Active"}
    ${"Completed"}
    `("Should filter and display $filter todos", async ({filter}) => {
        render(
            <TodoPage/>
        )

        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a rainbow{Enter}");
        expect(await screen.findByText("eat a rainbow")).toBeInTheDocument();
        await userEvent.type(await screen.findByPlaceholderText("What's next ?"), "eat a burger{Enter}");
        expect(await screen.findByText("eat a burger")).toBeInTheDocument();
        const lists = await screen.findAllByRole("list");
        const chkboxs = await within(lists[0]).findAllByRole("checkbox");
        await userEvent.click(chkboxs[0]);

        await userEvent.click(await screen.findByRole("link", {name: filter}));
        expect(lists[0].childElementCount).toEqual(filter === "All" ? 2 : 1);
        expect(await screen.findByText(filter === "Active" ? "eat a burger" : "eat a rainbow")).toBeInTheDocument();
    });

});