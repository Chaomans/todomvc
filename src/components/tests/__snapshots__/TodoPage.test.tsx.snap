// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoPage component Should render properly 1`] = `
<DocumentFragment>
  <section
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What's next ?"
        type="text"
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as completed
      </label>
    </section>
  </section>
</DocumentFragment>
`;
